### PR TITLE
Text to speech

### PR DESCRIPTION
**Pull request by TiviPlus**
## About The Pull Request

https://github.com/tgstation/tgstation/pull/74775/files
https://github.com/tgstation/tgstation/pull/75501

https://github.com/tgstation/tgstation/pull/75557
https://github.com/tgstation/tgstation/pull/75556

https://github.com/tgstation/tgstation/pull/75559/files

60 ish voiceds to choose from robots should sound like robots, pref toggle
Will need me to get config from mso

## Why It's Good For The Game
Its not >:)

## Changelog
:cl:
add: Added text so speech
/:cl:

### CODE REVIEW

**Comment by Three333Digit** [comment]
IT'S HERE!!!!!!!!

**Comment by novaepee** [comment]
image

**Comment by novaepee** [comment]
the-voices-voices-in-my-head

**Comment by zzurge** [comment]
Give me the plant

**Comment by Araulius** [comment]
inb4 someone uses this to shout the gamer word.

**Comment by FosterGrime-Heymynameslouie** [comment]
clown-clowntoclown

**Comment by Aroliacue** [comment]
FINALLY, OORAH

**Comment by novaepee** [comment]
No TTS on xenomorphs and EORD humans. Need to check if ERT get TTS.

**Comment by Aporhtonoma** [comment]
the single greatest addition that has ever been implemented in tgmc

**Comment by RipGrayson** [comment]
This is quite possibly the funniest thing I've ever seen on TGMC 10/10

**Comment by BlueHNT** [comment]
It causes immense delays.. Even if it's funny, that fun is costly.

**Comment by lKiseki** [comment]
Ever since the addition of TTS as a test merge, there has been a sharp increase in game lag and time dilation. Although I like the fun related to the voices and the memes, I do not believe that the sheer amount of lag that this provides (and it is by no means a small amount) is worth it, especially in TGMC -- we have a fast-paced, relatively competitive environment, and this lag is really causing an impact.

I would at least like to see TTS held off until such time as more upgrades can be done to the server; the lag is far too great for me to personally consider this to be worth it.

**Comment by Watermelon914** [comment]
Text to speech isn't going to have any significant performance impact to the server.
The only performance impact that TTS might have is with client IO, which specifically means that if you have bad latency, your game might freeze up when receiving the TTS audio files but this only really happens to a small selection of players. You can fix this by turning TTS off for yourself in the preferences menu.

**Comment by ProfessorPopoff** [comment]
Ladies, Gentlemen, anything in-between and off to the side...

This is it. 

It's over.

File: code/modules/client/preferences_ui.dm [hunk 1]
```diff
@@ -496,6 +498,23 @@
...
+				voices = json_decode(text_data)
+			if(!length(voices))
+				return
+			var/choice = tgui_input_list(ui.user, "What do you sound like?", "TTS", voices)
+			if(choice)
+				tts_voice = choice
+				if(TIMER_COOLDOWN_CHECK(user, COOLDOWN_TRY_TTS))
+					return
+				TIMER_COOLDOWN_START(ui.user, COOLDOWN_TRY_TTS, 0.5 SECONDS)
+				INVOKE_ASYNC(SStts, TYPE_PROC_REF(/datum/controller/subsystem/tts, queue_tts_message), ui.user.client, "Hello, this is my voice.", speaker = choice, local = TRUE)
```
Comment (Lumipharon) [inline code/modules/client/preferences_ui.dm:516]:
```suggestion
			if(!choice)
				return
			tts_voice = choice
			if(TIMER_COOLDOWN_CHECK(user, COOLDOWN_TRY_TTS))
				return
			TIMER_COOLDOWN_START(ui.user, COOLDOWN_TRY_TTS, 0.5 SECONDS)
			INVOKE_ASYNC(SStts, TYPE_PROC_REF(/datum/controller/subsystem/tts, queue_tts_message), ui.user.client, "Hello, this is my voice.", speaker = choice, local = TRUE)
```

**Review by Lumipharon — CHANGES_REQUESTED** [review]
need an early return

**Comment by grinkgo** [comment]
is there a less cringe way to preview all the voices

**Comment by CCorvet** [comment]
Some voices should probably the removed, like the child voices
Also yeah, some different way to check how the voices sound would be neat.

**Review by Lumipharon — APPROVED** [review]
We have reached the peak. It's all downhill from here.

**Pull request merged**